### PR TITLE
docs: Fixed Querier documentation for deduplication algorithm; Added FAQ

### PR DESCRIPTION
**Pull request by bwplotka**
Fixes: https://github.com/thanos-io/thanos/issues/5279

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

File: docs/components/query.md [hunk 1]
```diff
@@ -86,22 +114,71 @@ WITHOUT this replica flag (deduplication turned off), we will get 3 results:
...
+
+#### Penalty Algorithm
+
+Current algorithm Querier (and Compactor) uses for deduplication process can be called `penalty based`. The implementation is available [here](https://github.com/thanos-io/thanos/blob/d218e605c2dddab041551278bc1a5f632640f791/pkg/dedup/iter.go#L39). Generally it uses "the most stable" iterator (so source of series, e.g. Prometheus replica). The stability is detected by taking the first sample from both iterators and assigning penalty of 5s to the slower. Now if stable replica suddenly gives no sample, or sample with older timestamps plus penalty we switch to second one and give penalty for first one. This allows smoothly switching to stable replicas and still assuring the most consistent sample frequency possible.
+
+In details:
+
+Before deduplication does anything we remove exactly the same TSDB chunks from StoreAPI results [here](https://github.com/thanos-io/thanos/blob/de0e3848ff6085acf89a5f77e053c555a2cce550/pkg/query/iter.go#L76).
+
+Then if deduplication and replica labels were configured `dedup.NewSeriesSet` is invoked. It's algorithm can be outlined as follows:
```
Comment (jessicalins) [inline docs/components/query.md:171]:
```suggestion
If deduplication and replica labels were configured then `dedup.NewSeriesSet` is invoked. It's algorithm can be outlined as follows:
```

**Review by jessicalins — COMMENTED** [review]
cool, left some suggestions :)

File: docs/components/query.md [hunk 2]
```diff
@@ -1,65 +1,93 @@
...
+* Expose multiple [ExemplarAPI](https://github.com/thanos-io/thanos/blob/8fdd90c34318f453e0fda9b16ab2b201ed2ee4aa/pkg/exemplars/exemplarspb/rpc.proto#L25) gRPC backends as standard [Query Exemplars API](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-exemplars)
+* Expose multiple [TargetAPI](https://github.com/thanos-io/thanos/blob/f5174e098b30413501a6c930769e57228c13aa68/pkg/targets/targetspb/rpc.proto#L26) gRPC backends as standard [Targets API](https://prometheus.io/docs/prometheus/latest/querying/api/#targets)
+* Expose multiple [RulesAPI](https://github.com/thanos-io/thanos/blob/435700296491133dda9a2978bab46d0ba2e4e4f3/pkg/rules/rulespb/rpc.proto#L26) gRPC backends as standard [Rules API](https://prometheus.io/docs/prometheus/latest/querying/api/#rules)
+* Expose multiple [MetadataAPI](https://github.com/thanos-io/thanos/blob/af4ee3a09f6b00d280f5b6913b18864a595dc96f/pkg/metadata/metadatapb/rpc.proto#L23) gRPC backends as standard [Metric Metadata API](https://prometheus.io/docs/prometheus/latest/querying/api/#querying-metric-metadata)
 
-### Global View
+> Are you missing an API that you might find useful to use? Let us know on GitHub Issues!
```
[the remark this answers is not in the document]
Comment (bwplotka) [reply docs/components/query.md:42]: Good, point, will add in later commit or PR, since this is all green already

File: docs/components/query.md [hunk 3]
```diff
@@ -1,65 +1,93 @@
...
-Querier is fully stateless and horizontally scalable.
+Apart from federating PromQL based Query APIs, Querier is capable of federating other metric specific endpoints:
+
+* Labels
+* Exemplars
+* Scrape Targets
+* Alerts and Rules
+* Metadata
+
+> Querier is fully stateless and horizontally scalable. Querier is designed to work well no matter if you use sidecar, receiver or mixed Thanos deployment.
```
Comment (PhilipGough) [inline docs/components/query.md:15]:
```suggestion
> Querier is fully stateless and horizontally scalable. Querier is designed to work equally well with sidecar, receiver or mixed Thanos deployment.
```

File: docs/components/query.md [hunk 4]
```diff
@@ -1,65 +1,93 @@
...
+    --rule             "<rule-api>:<grpc-port>" \
+    --exemplar         "<exemplar-api>:<grpc-port>"
 ```
 
-## Querier use cases, why do I need this component?
+## Querier Use Cases
+
+> Why do I need this component?
+
+Thanos Querier allows to aggregate and optionally deduplicate multiple data backends under single endpoint.
```
Comment (PhilipGough) [inline docs/components/query.md:32]:
```suggestion
Thanos Querier allows to aggregate and optionally deduplicate multiple data backends under a single endpoint.
```

File: docs/components/query.md [hunk 5]
```diff
@@ -1,65 +1,93 @@
...
 
-### Run-time deduplication of HA groups
+### Use Case 2: Run-time deduplication of HA groups
 
-Prometheus is stateful and does not allow replicating its database. This means that increasing high availability by running multiple Prometheus replicas is not very easy to use. Simple load balancing will not work as for example after some crash, replica might be up but querying such replica will result in small gap during the period it was down. You have a second replica that maybe was up, but it could be down in other moment (e.g rolling restart), so load balancing on top of those is not working well.
+Prometheus is stateful and by design does not allow replicating its database. This means that increasing high availability by running multiple Prometheus replicas is not very easy to use. Simple load balancing will not work as for example after some crash, a replica might be up but querying such replica will result in small gap during the period it was down. You have a second replica that maybe was up, but it could be down in other moment (e.g. rolling restart), so requests based load-balancing on top of those two or more instances won't be accurate.
```
Comment (PhilipGough) [inline docs/components/query.md:68]: > This means that increasing high availability by running multiple Prometheus replicas is not very easy to use

Not very easy to use in what respect? Perhaps we can reword this to something like  

```
 well.
Prometheus is stateful and by design does not allow replicating its database. Therefore increasing high availability by running multiple Prometheus replicas can introduce some challenges from a querying perspective.
```

File: docs/components/query.md [hunk 6]
```diff
@@ -1,65 +1,93 @@
...
 
-Thanos Querier instead pulls the data from both replicas, and deduplicate those signals, filling the gaps if any, transparently to the Querier consumer.
+To solve this Thanos Querier pulls the data from both replicas, and deduplicate those signals, filling the gaps if any, transparently to the Querier consumer. Read more about deduplication process in [Deduplication](#deduplication) section.
```
[the remark this answers is not in the document]
Comment (PhilipGough) [reply docs/components/query.md:70]:
```suggestion
To solve this, Thanos Querier pulls the data from both replicas, and deduplicates those signals whilst filling any gaps transparently to the Querier consumer. Read more about the deduplication process in [Deduplication](#deduplication) section.
```

File: docs/components/query.md [hunk 7]
```diff
@@ -1,65 +1,93 @@
...
+The query layer can deduplicate series that were collected from high-availability pairs of data sources such as Prometheus or Receivers. A fixed single or multiple replica labels has to be configured at the start of Querier. Those special labels will be removed from each merged series. If, after label removal the series are duplicated, the deduplicate process occurs. Two or more series that are only distinguished by the given replica label, will be merged into a single time series.
+
+In practice the deduplication algorithm assumes that the same data was collected in the same or different times. The goal of this algorithm is to provide roughly the most consistent sample interval and only use the duplicated series as a fallback when main series is not following it's normal interval. This allows hiding gaps in collection of a single data source (e.g Prometheus) if another instance have data for it. It also works well to hide storage 1:1 replication done by Receivers. More on that later.
```
Comment (PhilipGough) [inline docs/components/query.md:84]:
```suggestion
In practice, the deduplication algorithm assumes that the same data was collected in the same or different times. The goal of this algorithm is to provide roughly the most consistent sample interval and only use the duplicated series as a fallback when main series is not following its normal interval. This allows hiding gaps in collection of a single data source (e.g Prometheus) if another instance has data for it. It also works well to hide storage 1:1 replication done by Receivers. More on that later.
```

File: docs/faq.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# Frequently Asked Questions
+
+## Querier
+
+#### How does Thanos Query show the user the same metric values from the two Prometheus instances?
+
```
Comment (PhilipGough) [inline docs/faq.md:6]: What do you think about the following? I'd also wonder should we expand on the point about the availability of the object storage?

```suggestion
Querier supports this via two different modes of deduplication!

Online deduplication - Querier transparently deduplicates metrics from replicated sources such as Prometheus replicas, replicated ingestion via Receivers and more. You can read more about online deduplication in the [Querier component docs](components/query.md).

Offline deduplication - This occurs inside [Compactor Vertical Compaction](components/compact.md#vertical-compactions) logic which offers deduplication of data at rest. This can be used when you want to reduce your storage size (make sure your object storage is available enough you can afford no back up of your data!). When offline deduplication is enabled, Querier will have no duplicated data to deduplicate.
```

**Review by PhilipGough — CHANGES_REQUESTED** [review]
Some comments but looking great overall!

File: docs/components/query.md (hunk 5 above)
Comment (bwplotka) [reply docs/components/query.md:68]: sgtm!

File: docs/faq.md (hunk 8 above)
Comment (bwplotka) [reply docs/faq.md:6]: Ack, will take a look this week
